src/main/java/org/mockito/internal/creation/bytebuddy/SubclassBytecodeGenerator.java

```diff
@@ -90,15 +90,53 @@ protected SubclassBytecodeGenerator(
         handler = ModuleHandler.make(byteBuddy, loader, random);
     }
 
+    private static boolean needsSamePackageClassLoader(MockFeatures<?> features) {
+        if (!Modifier.isPublic(features.mockedType.getModifiers())
+                || !features.mockedType.isInterface()) {
+            // The mocked type is package private or is not an interface and thus may contain
+            // package
+            // private methods.
+            return true;
+        }
+        for (Class<?> iface : features.interfaces) {
+            if (!Modifier.isPublic(iface.getModifiers())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
     @Override
     public <T> Class<? extends T> mockClass(MockFeatures<T> features) {
-        ClassLoader classLoader =
+        MultipleParentClassLoader.Builder loaderBuilder =
                 new MultipleParentClassLoader.Builder()
-                        .appendMostSpecific(getAllTypes(features.mockedType))
-                        .appendMostSpecific(features.interfaces)
-                        .appendMostSpecific(currentThread().getContextClassLoader())
-                        .appendMostSpecific(MockAccess.class)
-                        .build();
+                        .append(features.mockedType)
+                        .append(features.interfaces)
+                        .append(MockAccess.class, DispatcherDefaultingToRealMethod.class)
+                        .append(
+                                MockMethodInterceptor.class,
+                                MockMethodInterceptor.ForHashCode.class,
+                                MockMethodInterceptor.ForEquals.class);
+        ClassLoader contextLoader = currentThread().getContextClassLoader();
```

Comment:
I'd prefer to avoid the static import here as it's the only use and we tend to not use static imports.

Reply:
We already did this on line 99 in the current implementation, so we don't change this here. Can we clean these up in a follow-up?

Reply:
Let's figure this out in a follow-up. I think we need to reconfigure the formatter to do this.

Reply:
Agreed.